File: api/cliente_ejemplo.php
<?php
declare(strict_types=1);

/**
 * jsonSQLDB - Cliente de ejemplo.
 *
 * Copia este fichero en la aplicación que vaya a consumir la API. Solo necesita
 * la URL del endpoint, su API key y el secreto HMAC.
 *
 *   $cli = JsonSqlDbCliente::pruebas();
 *   $filas = $cli->consultar('SELECT * FROM clientes WHERE ciudad = ?', ['Torrevieja']);
 *
 * O montándolo a mano, que es lo que hará tu aplicación de verdad:
 *
 *   $cli = new JsonSqlDbCliente('https://miservidor/jsonsqldb/api/jsonsqldb_api.php',
 *                               'MI_API_KEY', 'MI_HMAC_SECRET', 'mibase');
 *
 * Si la API va por HTTPS con certificado propio:
 *   $cli->certificado('C:/xampp/apache/conf/ssl.crt/server.crt');
 *   $cli->aceptarAutofirmado();   // atajo, sin verificar
 */
final class JsonSqlDbCliente
{
    private string $url;
    private string $apiKey;
    private string $secreto;
    private string $base;
    private int    $timeout;
    private string $ca          = '';
    private bool   $autofirmado = false;

    // Datos de la clave de los ejemplos: permiso de escritura sobre 'pruebas'.
    // La misma que usa cliente_ejemplo.ps1. Para tu aplicación, crea una clave
    // propia en api/jsonsqldb_api_config.php.
    public const EJEMPLO_URL     = 'https://example.com/jsonsqldb/api/jsonsqldb_api.php';
    public const EJEMPLO_API_KEY = 'CHANGE_ME_EXAMPLE_API_KEY';
    public const EJEMPLO_SECRETO = 'CHANGE_ME_EXAMPLE_SECRET';
    public const EJEMPLO_BASE    = 'pruebas';

    /** Cliente ya configurado contra la base 'pruebas'. */
    public static function pruebas(): self
    {
        $cli = new self(self::EJEMPLO_URL, self::EJEMPLO_API_KEY, self::EJEMPLO_SECRETO, self::EJEMPLO_BASE);
        $cli->aceptarAutofirmado();
        return $cli;
    }

    public function __construct(string $url, string $apiKey, string $secreto, string $base, int $timeout = 60)
    {
        $this->url     = $url;
        $this->apiKey  = $apiKey;
        $this->secreto = $secreto;
        $this->base    = $base;
        $this->timeout = $timeout;
    }

    /**
     * Certificado del servidor de la API cuando es propio o de una CA interna.
     * Se sigue verificando, pero contra este fichero .crt / .pem.
     *
     *   $cli->certificado('C:/xampp/apache/conf/ssl.crt/server.crt');
     */
    public function certificado(string $rutaCrt): self
    {
        if (!is_file($rutaCrt) || !is_readable($rutaCrt)) {
            throw new RuntimeException("No se puede leer el certificado: $rutaCrt");
        }
        $this->ca = $rutaCrt;
        return $this;
    }

    /**
     * Aceptar el certificado sin comprobarlo (autofirmado en red interna).
     * Es peor que certificado(): deja de protegerte frente a un intermediario.
     */
    public function aceptarAutofirmado(bool $si = true): self
    {
        $this->autofirmado = $si;
        return $this;
    }

    /**
     * Ejecuta una sentencia. Cada ? de la SQL toma su valor de $parametros, que
     * viajan aparte y el servidor inserta ya analizados: nunca se concatenan al
     * texto de la consulta, así que un valor no puede alterarla.
     *
     * Valores admitidos: null, bool, int, float y string.
     *
     * @return array filas del SELECT, o ['success'=>true,'filas'=>n,'mensaje'=>'...']
     * @throws RuntimeException si la API devuelve un error
     */
    public function consultar(string $sql, array $parametros = []): array
    {
        $params = $parametros === []
            ? ''
            : (string)json_encode(array_values($parametros), JSON_UNESCAPED_UNICODE);

        $timestamp = (string)time();
        $token     = hash_hmac(
            'sha256',
            '+' . $this->apiKey . '|' . $timestamp . '|' . $sql . $params . '¿',
            $this->secreto
        );

        $post = http_build_query([
            'api_key'   => $this->apiKey,
            'db'        => $this->base,
            'sql'       => $sql,
            'params'    => $params,
            'timestamp' => $timestamp,
            'token'     => $token,
        ]);

        $verificar = $this->ca !== '' || !$this->autofirmado;

        $ch = curl_init($this->url);
        $opciones = [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => $post,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => $this->timeout,
            CURLOPT_SSL_VERIFYPEER => $verificar,
            CURLOPT_SSL_VERIFYHOST => $verificar ? 2 : 0,
        ];
        if ($this->ca !== '') {
            $opciones[CURLOPT_CAINFO] = $this->ca;
        }
        curl_setopt_array($ch, $opciones);
        $respuesta = curl_exec($ch);
        $errorCurl = curl_error($ch);
        curl_close($ch);

        if ($respuesta === false) {
            throw new RuntimeException('No se pudo conectar con la API: ' . $errorCurl);
        }

        $datos = json_decode($respuesta, true);
        if (!is_array($datos)) {
            throw new RuntimeException('Respuesta no válida de la API: ' . substr((string)$respuesta, 0, 200));
        }
        if (isset($datos['error'])) {
            throw new RuntimeException((string)$datos['error']);
        }
        return $datos;
    }
}

File: api/cliente_ejemplo.py
"""
jsonSQLDB - Cliente de ejemplo en Python.

Copia este fichero en la aplicación que vaya a consumir la API. Solo usa la
biblioteca estándar: sin pip, sin requests, sin dependencias.

    from cliente_ejemplo import JsonSqlDbCliente

    cli = JsonSqlDbCliente(
        "https://miservidor/jsonsqldb/api/jsonsqldb_api.php",
        "MI_API_KEY", "EL_HMAC_SECRET_DE_ESA_KEY", "mibase")

    filas = cli.consultar("SELECT * FROM clientes WHERE ciudad = ?", ["Madrid"])

EL SECRETO ES EL DE LA CLAVE: el campo 'hmac_secret' de esa misma entrada de
$API_KEYS en api/jsonsqldb_api_config.php. Cada cuenta tiene el suyo, distinto
del de las demás.

Los valores NO se concatenan a la SQL: se ponen ? y los valores van aparte. El
servidor los inserta ya analizados en el árbol de la consulta, así que un valor
no puede alterarla por mucho SQL que contenga.

Requiere Python 3.7 o superior.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import ssl
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, List, Optional, Sequence, Union

__all__ = ["JsonSqlDbCliente", "JsonSqlDbError"]

Fila = Dict[str, Any]
Resultado = Union[List[Fila], Dict[str, Any]]


class JsonSqlDbError(RuntimeError):
    """Error devuelto por la API, o fallo al llamarla."""


class JsonSqlDbCliente:
    def __init__(
        self,
        url: str,
        api_key: str,
        hmac_secret: str,
        base: str,
        timeout: int = 60,
    ) -> None:
        self.url = url
        self.api_key = api_key
        self.hmac_secret = hmac_secret   # el 'hmac_secret' de esta API key
        self.base = base
        self.timeout = timeout
        self._ca: Optional[str] = None
        self._autofirmado = False

    # ------------------------------------------------------------------
    # Certificado del servidor
    # ------------------------------------------------------------------

    def certificado(self, ruta_crt: str) -> "JsonSqlDbCliente":
        """
        Certificado del servidor cuando es propio o de una CA interna.
        Se sigue verificando, pero contra este fichero .crt / .pem.
        """
        self._ca = ruta_crt
        return self

    def aceptar_autofirmado(self, si: bool = True) -> "JsonSqlDbCliente":
        """
        Aceptar el certificado sin comprobarlo. Vale en una red interna de
        confianza, pero deja de protegerte frente a un intermediario: es peor
        que certificado().
        """
        self._autofirmado = si
        return self

    # ------------------------------------------------------------------
    # Consulta
    # ------------------------------------------------------------------

    def consultar(self, sql: str, parametros: Sequence[Any] = ()) -> Resultado:
        """
        Ejecuta una sentencia.

        Devuelve la lista de filas en SELECT y SHOW, o un diccionario
        {'success': True, 'filas': n, 'mensaje': '...'} en las escrituras.

        Valores admitidos en los parámetros: None, bool, int, float y str.
        """
        params = ""
        if parametros:
            # separators sin espacios: el servidor firma el texto EXACTO que
            # recibe, así que lo que se firma y lo que se envía deben coincidir
            params = json.dumps(
                list(parametros), ensure_ascii=False, separators=(",", ":")
            )

        timestamp = str(int(time.time()))
        mensaje = f"+{self.api_key}|{timestamp}|{sql}{params}\u00bf"
        token = hmac.new(
            self.hmac_secret.encode("utf-8"),
            mensaje.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()

        cuerpo = urllib.parse.urlencode(
            {
                "api_key": self.api_key,
                "db": self.base,
                "sql": sql,
                "params": params,
                "timestamp": timestamp,
                "token": token,
            }
        ).encode("utf-8")

        datos = self._enviar(cuerpo)

        if isinstance(datos, dict) and "error" in datos:
            raise JsonSqlDbError(str(datos["error"]))
        return datos

    def valor(self, sql: str, parametros: Sequence[Any] = ()) -> Any:
        """Primer valor de la primera fila, o None si no hay filas."""
        filas = self.consultar(sql, parametros)
        if isinstance(filas, list) and filas:
            return next(iter(filas[0].values()), None)
        return None

    def bases(self) -> List[str]:
        """Lista de bases de datos. Necesita una API key con acceso a todas."""
        anterior, self.base = self.base, ""
        try:
            filas = self.consultar("SHOW DATABASES")
        finally:
            self.base = anterior
        return [f["base"] for f in filas] if isinstance(filas, list) else []

    # ------------------------------------------------------------------
    # Interno
    # ------------------------------------------------------------------

    def _enviar(self, cuerpo: bytes) -> Resultado:
        peticion = urllib.request.Request(
            self.url,
            data=cuerpo,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
            method="POST",
        )

        contexto: Optional[ssl.SSLContext] = None
        if self.url.lower().startswith("https://"):
            if self._ca:
                contexto = ssl.create_default_context(cafile=self._ca)
            elif self._autofirmado:
                contexto = ssl.create_default_context()
                contexto.check_hostname = False
                contexto.verify_mode = ssl.CERT_NONE

        try:
            with urllib.request.urlopen(
                peticion, timeout=self.timeout, context=contexto
            ) as respuesta:
                texto = respuesta.read().decode("utf-8")
        except urllib.error.HTTPError as e:
            # La API responde con JSON incluso en los códigos de error
            texto = e.read().decode("utf-8", errors="replace")
        except (urllib.error.URLError, OSError) as e:
            raise JsonSqlDbError(f"No se pudo llamar a la API ({self.url}): {e}") from e

        try:
            return json.loads(texto)
        except json.JSONDecodeError as e:
            raise JsonSqlDbError(
                f"Respuesta no válida de la API: {texto[:300]}"
            ) from e


# ----------------------------------------------------------------------
# Ejemplos
# ----------------------------------------------------------------------

if __name__ == "__main__":
    cli = JsonSqlDbCliente(
        "https://example.com/jsonsqldb/api/jsonsqldb_api.php",
        "CHANGE_ME_EXAMPLE_API_KEY",
        "CHANGE_ME_EXAMPLE_SECRET",
        "pruebas",
    )

    # Certificado propio o autofirmado, si hace falta:
    # cli.certificado("C:/xampp/apache/conf/ssl.crt/server.crt")
    # cli.aceptar_autofirmado()

    # --- Consultas ---
    print(cli.consultar("SHOW TABLES"))
    print(cli.consultar("SELECT * FROM clientes LIMIT 10"))

    # Con parámetros: el valor viaja aparte, nunca dentro de la SQL
    print(cli.consultar("SELECT * FROM clientes WHERE nombre = ?", ["O'Donnell"]))

    # Un valor con SQL dentro se busca literalmente: no altera la consulta
    print(
        cli.consultar(
            "SELECT * FROM clientes WHERE nombre = ?",
            ["x'); DROP TABLE clientes; --"],
        )
    )

    # IN de tamaño variable: un ? por elemento
    ciudades = ["Torrevieja", "Alicante", "Murcia"]
    huecos = ",".join("?" for _ in ciudades)
    print(cli.consultar(f"SELECT nombre FROM clientes WHERE ciudad IN ({huecos})", ciudades))

    # --- Escrituras ---
    # cli.consultar("INSERT INTO clientes (cod, nombre, saldo) VALUES (?, ?, ?)",
    #               ["A1", "Ana", 10.55])
    # cli.consultar("UPDATE clientes SET saldo = saldo + ? WHERE id = ?", [25.40, 7])
    # cli.consultar("DELETE FROM clientes WHERE id = ?", [7])

    # --- Bases de datos: db vacío ---
    # print(cli.bases())

    # --- Volcar a CSV ---
    # import csv
    # filas = cli.consultar("SELECT * FROM clientes")
    # if filas:
    #     with open("clientes.csv", "w", newline="", encoding="utf-8-sig") as f:
    #         w = csv.DictWriter(f, fieldnames=list(filas[0].keys()), delimiter=";")
    #         w.writeheader()
    #         w.writerows(filas)
